include/dali/core/access_order.h

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2021-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_ACCESS_ORDER_H_
+#define DALI_CORE_ACCESS_ORDER_H_
+
+#include <cuda_runtime.h>
+#include <cstddef>
+#include "dali/core/api_helper.h"
+
+namespace dali {
+
+/**
+ * @brief Describes the order in which an object is accessed. It can be stream or host.
+ *
+ * The access order can be either a CUDA stream order or host order. There's also
+ * a special "null" value, useful when passing AccessOrder to functions/objects which may have
+ * internal streams.
+ */
+class DLL_PUBLIC AccessOrder {
+ public:
+  constexpr AccessOrder() = default;
+  constexpr AccessOrder(cudaStream_t stream, int device_id)
+  : stream_(stream), device_id_(device_id) {}
+
+  AccessOrder(int) = delete;  // NOLINT  prevent construction from 0
+  AccessOrder(std::nullptr_t) = delete;  // NOLINT
+
+  AccessOrder(cudaStream_t stream);  // NOLINT
+
+  static constexpr AccessOrder host() {
+    return AccessOrder(host_sync_stream(), -1);
+  }
+
+  /**
+   * @brief Returns the underlying handle.
+   *
+   * Returns the underlying handle. It can be either a genuine CUDA stream or a
+   * special host-sync value. When there's no value, the return value is 0, but it should
+   * not be treated as a valid CUDA stream.
+   */
+  cudaStream_t get() const noexcept { return has_value() ? stream_ : 0; }
+
+  /**
+   * @brief Returns stream handle, if any.
+   *
+   * Returns the underlying handle when it denotes a CUDA stream. If this order does not
+   * represent a valid CUDA stream, the return value is 0, but it should not be treated
+   * as a valid CUDA stream. This behaviour is to facilitate the interoperatibility with
+   * legacy functions that take cudaStream_t.
```

Comment:
```suggestion
   * legacy functions that take cudaStream_t. Use get() when possible.
```

Reply:
It's not that simple :) In general, when passing a value to a function that has no notion of our special host pseudostream, it's safer to use `stream()` than `get()`. The legacy part more about functions that don't need a stream, but take one nonetheless.

Reply:
Understood.